consensus/application.go

```diff
@@ -608,119 +644,88 @@ func (ms *MidState) ApplyBlock(b types.Block, bs V1BlockSupplement) {
+	for i := range sces {
+		sce := &sces[i]
+		fn(siacoinLeaf(&sce.SiacoinElement, sce.Spent))
 	}
-	for i := range ms.sfes {
-		fn(siafundLeaf(&ms.sfes[i], ms.isSpent(ms.sfes[i].ID)))
+	for i := range sfes {
+		sfe := &sfes[i]
+		fn(siafundLeaf(&sfe.SiafundElement, sfe.Spent))
 	}
-	for i := range ms.fces {
-		if r, ok := ms.revs[ms.fces[i].ID]; ok {
-			fn(fileContractLeaf(r, ms.isSpent(ms.fces[i].ID)))
-		} else {
-			fn(fileContractLeaf(&ms.fces[i], ms.isSpent(ms.fces[i].ID)))
-		}
+	for i := range fces {
+		fce := &fces[i]
+		fn(fileContractLeaf(&fce.FileContractElement, fce.Revision, fce.Resolved))
 	}
-	for i := range ms.v2fces {
-		if r, ok := ms.v2revs[ms.v2fces[i].ID]; ok {
-			fn(v2FileContractLeaf(r, ms.isSpent(ms.v2fces[i].ID)))
-		} else {
-			fn(v2FileContractLeaf(&ms.v2fces[i], ms.isSpent(ms.v2fces[i].ID)))
-		}
+	for i := range v2fces {
+		v2fce := &v2fces[i]
+		fn(v2FileContractLeaf(&v2fce.V2FileContractElement, v2fce.Revision, v2fce.Resolution != nil))
 		// NOTE: Although it is an element, we do not process the ProofIndex
 		// field of V2StorageProofs. These are a special case, as they are not
 		// being updated (like e.g. siacoin inputs), nor are they being created
 		// (like e.g. attestations). In other words, they have no effect on the
 		// accumulator, and thus including them would only cause confusion.
 	}
-	for i := range ms.aes {
-		fn(attestationLeaf(&ms.aes[i]))
+	for i := range aes {
+		fn(attestationLeaf(&aes[i]))
 	}
-	fn(chainIndexLeaf(&ms.cie))
+	fn(chainIndexLeaf(cie))
 }
 
-func (ms *MidState) forEachRevertedElement(fn func(elementLeaf)) {
-	for i := range ms.sces {
-		fn(siacoinLeaf(&ms.sces[i], false))
+func forEachRevertedElement(sces []SiacoinElementDiff, sfes []SiafundElementDiff, fces []FileContractElementDiff, v2fces []V2FileContractElementDiff, fn func(elementLeaf)) {
+	for i := range sces {
+		fn(siacoinLeaf(&sces[i].SiacoinElement, false))
 	}
-	for i := range ms.sfes {
-		fn(siafundLeaf(&ms.sfes[i], false))
+	for i := range sfes {
+		fn(siafundLeaf(&sfes[i].SiafundElement, false))
 	}
-	for i := range ms.fces {
-		fn(fileContractLeaf(&ms.fces[i], false))
+	for i := range fces {
+		fn(fileContractLeaf(&fces[i].FileContractElement, nil, false))
 	}
-	for i := range ms.v2fces {
-		fn(v2FileContractLeaf(&ms.v2fces[i], false))
+	for i := range v2fces {
+		fn(v2FileContractLeaf(&v2fces[i].V2FileContractElement, nil, false))
 	}
 }
 
 // An ApplyUpdate represents the effects of applying a block to a state.
 type ApplyUpdate struct {
-	ms  *MidState
+	sces   []SiacoinElementDiff
+	sfes   []SiafundElementDiff
+	fces   []FileContractElementDiff
+	v2fces []V2FileContractElementDiff
+	aes    []types.AttestationElement
+	cie    types.ChainIndexElement
+
 	eau elementApplyUpdate
 }
 
+// SiacoinElementDiffs returns the siacoin element diffs related to the applied
+// block.
+func (au ApplyUpdate) SiacoinElementDiffs() []SiacoinElementDiff { return au.sces }
```

Comment:
Is there a benefit to wrapping the field access in a method rather than exposing the slice directly?

Reply:
I considered it, but methods let you define an `Update` interface, and give us some flexibility if we end up changing the internal representation again